components/feature/addons/src/main/java/mozilla/components/feature/addons/Addon.kt

```diff
@@ -193,8 +194,118 @@ data class Addon(
          * @param permissions The list of permissions to be localized. Valid permissions can be found in
          * https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#API_permissions
          */
-        fun localizePermissions(permissions: List<String>): List<Int> {
-            return permissions.mapNotNull { permissionToTranslation[it] }
+        fun localizePermissions(permissions: List<String>, context: Context): List<String> {
+            var localizedUrlAccessPermissions = emptyList<String>()
+            val requireAllUrlsAccess = permissions.contains("<all_urls>")
+            val notFoundPermissions = mutableListOf<String>()
+
+            val localizedNormalPermissions = permissions.mapNotNull {
+                val id = permissionToTranslation[it]
+                if (id == null) notFoundPermissions.add(it)
+                id
+            }.map { context.getString(it) }
+
+            if (!requireAllUrlsAccess && notFoundPermissions.isNotEmpty()) {
+                localizedUrlAccessPermissions = localizedURLAccessPermissions(context, notFoundPermissions)
+            }
+
+            return localizedNormalPermissions + localizedUrlAccessPermissions
+        }
+
+        @Suppress("MaxLineLength")
+        internal fun localizedURLAccessPermissions(context: Context, accessPermissions: List<String>): List<String> {
+            val localizedSiteAccessPermissions = mutableListOf<String>()
+            val permissionsToTranslations = mutableMapOf<String, Int>()
+
+            accessPermissions.forEach { permission ->
+                val id = localizeURLAccessPermission(permission)
+                if (id != null) {
+                    permissionsToTranslations[permission] = id
+                }
+            }
+
+            if (permissionsToTranslations.values.any { it.isAllURLsPermission() }) {
+                localizedSiteAccessPermissions.add(context.getString(R.string.mozac_feature_addons_permissions_all_urls_description))
+            } else {
+                formatURLAccessPermission(permissionsToTranslations, localizedSiteAccessPermissions, context)
+            }
+            return localizedSiteAccessPermissions
+        }
+
+        @Suppress("MagicNumber", "ComplexMethod")
+        private fun formatURLAccessPermission(
+            permissionsToTranslations: MutableMap<String, Int>,
+            localizedSiteAccessPermissions: MutableList<String>,
+            context: Context
+        ) {
+            var domainCount = 0
+            var siteCount = 0
+
+            loop@ for ((permission, stringId) in permissionsToTranslations) {
+                var host = permission.toUri().host ?: ""
+                when {
+                    stringId.isDomainAccessPermission() -> {
+                        ++domainCount
+                        host = host.removePrefix("*.")
+
+                        if (domainCount > 4) continue@loop
```

Comment:
Maybe we should assign 4 to a private const val? so we can update quickly if we decided to switch to a different number.

Reply:
👍